src/ripple/app/paths/impl/DirectStep.cpp

```diff
@@ -33,7 +33,129 @@
 
 namespace ripple {
 
-class DirectStepI : public StepImp<IOUAmount, IOUAmount, DirectStepI>
+template <class Policy>
+class DirectStepI;
+
+// Flow is used in two different circumstances for transferring funds:
+//  o Payments, and
+//  o Offer crossing.
+// The rules for handling funds in these two cases are almost, but not
+// quite, the same.
+//
+// In order to make the differences between payments and offer crossing
+// apparent, those differences are relegated to two different policies:
+// one for payments and one for offer crossing.  Whichever policy is
+// appropriate is injected into the StepImp at construction.  This is a
+// very simple case of policy-based design with only one policy type.
+//
+// See: https://en.wikipedia.org/wiki/Policy-based_design
+//
+// The policy classes follow.
+
+// Class to inject into a DirectStepI for a payment (not offer crossing).
+class DirectIPaymentPolicy
+{
+public:
+    DirectIPaymentPolicy()
+    {
+    }
+
+    bool verifyPrevStepRedeems (bool) const
+    {
+        // A payment doesn't care whether or not prevStepRedeems.
+        return true;
+    }
+
+    bool verifyDstQualityIn (std::uint32_t dstQIn) const
+    {
+        // Payments have no particular expectations for what dstQIn will be.
+        return true;
+    }
+
+    // Return the quality for a given offer.
+    template <class Policy>
+    std::uint32_t
+    quality (DirectStepI<Policy> const& step,
+        ReadView const& sb,
+        // set true for quality in, false for quality out
+        bool qin) const;
+
+    // Compute the maximum value that can flow from src->dst at
+    // the best available quality.
+    // return: first element is max amount that can flow,
+    //         second is true if dst holds an iou from src.
+    //
+    // The definition uses DirectStepI, so it comes later in the file.
```

Comment:
Consider moving the declaration of policy classes after the step classes (I'd also remove the comment even if you keep the decl here).